test/extensions/tracers/opencensus/tracer_test.cc

```diff
@@ -0,0 +1,129 @@
+#include <iostream>
+#include <vector>
+
+#include "envoy/config/trace/v2/trace.pb.validate.h"
+
+#include "extensions/tracers/opencensus/opencensus_tracer_impl.h"
+
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/tracing/mocks.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "opencensus/trace/exporter/span_data.h"
+#include "opencensus/trace/exporter/span_exporter.h"
+#include "opencensus/trace/span_id.h"
+
+using testing::NiceMock;
+using testing::Return;
+
+namespace opencensus {
+namespace trace {
+namespace exporter {
+
+class SpanExporterTestPeer {
+public:
+  static constexpr auto& ExportForTesting = SpanExporter::ExportForTesting;
+};
+
+} // namespace exporter
+} // namespace trace
+} // namespace opencensus
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace OpenCensus {
+
+using ::opencensus::trace::exporter::SpanData;
+using ::opencensus::trace::exporter::SpanExporter;
+
+namespace {
+
+class SpanCatcher : public SpanExporter::Handler {
+public:
+  void Export(const std::vector<SpanData>& spans) override {
+    absl::MutexLock lock(&mu_);
+    for (const auto& span : spans) {
+      spans_.emplace_back(span);
+    }
+  }
+
+  // Force export, return generated spandata, and clear the catcher.
+  std::vector<SpanData> Catch() {
+    opencensus::trace::exporter::SpanExporterTestPeer::ExportForTesting();
+    absl::MutexLock lock(&mu_);
+    std::vector<SpanData> ret = std::move(spans_);
+    spans_.clear();
+    return ret;
```

Comment:
Not sure I fully grok what is running on which threads, can you add more detailed explanation of the threading behavior?

Reply:
Added comments.